components/sync_manager/src/manager.rs

```diff
@@ -0,0 +1,193 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+use crate::msg_types::{ServiceStatus, SyncParams, SyncResult};
+use logins::PasswordEngine;
+use places::{bookmark_sync::store::BookmarksStore, history_sync::store::HistoryStore, PlacesApi};
+use std::collections::HashMap;
+use std::sync::Mutex;
+use std::sync::{atomic::AtomicUsize, Arc, Weak};
+use sync15::MemoryCachedState;
+
+const LOGINS_ENGINE: &str = "passwords";
+const HISTORY_ENGINE: &str = "history";
+const BOOKMARKS_ENGINE: &str = "bookmarks";
+
+pub struct SyncManager {
+    mem_cached_state: Option<MemoryCachedState>,
+    places: Weak<PlacesApi>,
+    logins: Weak<Mutex<PasswordEngine>>,
+}
+
+impl SyncManager {
+    pub fn new() -> Self {
+        Self {
+            mem_cached_state: None,
+            places: Weak::new(),
+            logins: Weak::new(),
+        }
+    }
+
+    pub fn set_places(&mut self, places: Arc<PlacesApi>) {
+        self.places = Arc::downgrade(&places);
+    }
+
+    pub fn set_logins(&mut self, logins: Arc<Mutex<PasswordEngine>>) {
+        self.logins = Arc::downgrade(&logins);
+    }
+
+    pub fn disconnect(&mut self) {
+        if let Some(logins) = self
+            .logins
+            .upgrade()
+            .as_ref()
+            .map(|l| l.lock().expect("poisoned logins mutex"))
+        {
+            if let Err(e) = logins.reset() {
+                log::error!("Failed to reset logins: {}", e);
+            }
+        } else {
+            log::warn!("Unable to wipe logins, be sure to call set_logins before disconnect if this is surprising");
```

Comment:
"wipe" seems wrong here - IIUC, the reset() functions only delete metadata and not the bookmarks/history/logins themselves? (I guessed 'wipe' was a typo until I saw it repeated - it might still be one though :)